### PR TITLE
Add per-sample tool filtering to GRPOTrainer via `tools` column

### PR DESCRIPTION
**Pull request by lailanelkoussy**
# Add per-sample tool filtering to `GRPOTrainer`

## What does this PR do?

Adds **per-sample tool filtering** to `GRPOTrainer`. When the dataset contains a `"tools"` column,
each sample can specify which subset of the global `tools` pool is available, controlling both the
prompt schema (which tools the model sees) and execution (which tool calls are allowed to run).

### Motivation

In agentic GRPO training, different samples often require different tool subsets. For example, a math
problem should only expose a `calculator` tool, while a translation task should only expose a
`translator`. Without this feature, every sample sees the full tool list, leading to:

- **Noisy learning signal**: the model wastes exploration budget on irrelevant tools
- **Incorrect reward attribution**: tool call failures due to schema mismatch confuse the reward signal
- **No curriculum control**: no way to progressively introduce tools during training

### API

```python
from datasets import Dataset
from trl import GRPOTrainer

dataset = Dataset.from_dict({
    "prompt": [
        [{"role": "user", "content": "What is 2+2?"}],
        [{"role": "user", "content": "Translate 'hello'"}],
    ],
    "tools": [["calculator"], ["translator"]],  # ← per-sample tool column, auto-detected
})

trainer = GRPOTrainer(
    model=model,
    reward_funcs=reward,
    tools=[calculator, translator],
    train_dataset=dataset,
)
```

When the dataset has no `tools` column, behavior is **identical** to the existing API — full backward compatibility.

### Implementation details

| Area | Change |
|------|--------|
| **`_set_signature_columns_if_needed`** | Always includes `"tools"` so the column passes through the dataloader |
| **`_tokenize_prompts`** | New `per_sample_tools` param; tokenizes each prompt individually with its own tool schema when set |
| **`_tool_call_loop`** | New `per_sample_tools` param; builds per-sample sync/async tool dicts for execution filtering |
| **`_generate`** | Threads `per_sample_tools` to tokenization and tool-call loop |
| **`_generate_and_score_completions`** | Auto-detects `"tools"` key in batch; resolves tool names via per-sample `_sync_tool_dicts`/`_async_tool_dicts`; `None` entries fall back to the full tool list |

### Tests

New test file `tests/test_grpo_tools_column.py`:

**Validation tests:**
- Valid configuration (dataset with tools column) passes without error
- Backward compatibility: no tools column → trainer uses all tools
- Per-sample tool dict correctness
- Signature columns always include `"tools"`

**Integration tests:**
- End-to-end training with a `tools` dataset column (all samples, same tools)
- Training with tool subset restriction (some calls fail as expected)
- Backward compatibility (tools provided, no dataset tools column)
- `None` in tools column falls back to full tool list
- Async tools with per-sample filtering
- Environment factory with per-sample filtering (correct per-environment method binding)

### Files changed

- `trl/trainer/grpo_trainer.py` — core implementation
- `tests/test_grpo_tools_column.py` — new file (test suite)
- `docs/source/per_sample_tools.md` — new file (documentation)

## Before submitting

- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case)
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

---

> [!NOTE]
> **Medium Risk**
> Touches GRPO rollout/tokenization and tool-execution paths; incorrect filtering could change prompt schemas or tool dispatch during training despite backward-compat fallbacks and added tests.
> 
> **Overview**
> Adds **per-sample tool filtering** to `GRPOTrainer`: when the training dataset includes a `tools` column (list of tool `__name__` strings per row), the trainer now restricts both the rendered chat tool schema and runtime tool execution to that subset, with `None`/missing values falling back to the full global `tools` pool.
> 
> This threads a new `per_sample_tools` flow through prompt tokenization, generation, and the tool-call loop, and ensures `tools` is kept as a signature column so it isn’t dropped by column-pruning. Documentation is added (`docs/source/per_sample_tools.md` + toctree entry) and a comprehensive test suite is introduced covering validation, backward compatibility, subset behavior, `None` fallback, async tools, and interaction notes with `environment_factory`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 049b3002929a8c377b146ecd5dbc1583395c0ee4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -274,6 +295,7 @@ def __init__(
         optimizers: tuple[torch.optim.Optimizer | None, torch.optim.lr_scheduler.LambdaLR | None] = (None, None),
         peft_config: "PeftConfig | None" = None,
         tools: list[Callable] | None = None,
+        tools_column_name: str | None = None,
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: we don't need a new arg, just `tools` as name is enough

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -488,6 +510,30 @@ def __init__(
 
         self.tools = tools + (environment_methods[0] if self.environments is not None else [])
 
+        # Per-sample tool filtering support
+        self.tools_column_name = tools_column_name
+
+        # When tools_column_name is used, the registry is the sole source for resolving tool names from dataset rows.
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: I don't think this is useful, it's kinda obvious that we shouldn't have duplicates

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -625,6 +671,42 @@ def __init__(
 
         # Reference model
         self.beta = args.beta
+
+        # Validate per-sample tool names in the dataset(s)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: No this is too defensive as well

**Comment by lailanelkoussy** [comment]
Thanks for the feedback! I've simplified the implementation: removed the `tools_column_name` parameter entirely and the validation blocks. The trainer now just auto-detects a tools column in the dataset. Happy to adjust further if needed.

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1742,6 +1820,19 @@ def _generate_and_score_completions(
...
+        per_sample_tools = None
+        if self.tools and inputs and "tools" in inputs[0]:
+            tool_lookup = {tool.__name__: tool for tool in self.tools}
+            per_sample_tools = []
+            for example in inputs:
+                tool_names = example.get("tools")
+                if tool_names is not None:
+                    per_sample_tools.append([tool_lookup[name] for name in tool_names])
+                else:
+                    per_sample_tools.append(list(tool_lookup.values()))
```
[the remark this answers is not in the document]
Comment (lailanelkoussy) [reply trl/trainer/grpo_trainer.py:1838]: Acknowledged. `self.tools` uses `environment_methods[0]` because all environment instances share the same method signatures  (they're created by the same factory), so it's the right choice for schema rendering. For tool execution, the limitation you describe is real: per-sample filtering via the `tools` column resolves callables from `self.tools`, so all samples would execute on environment 0's instance. Combining `tools` column with `environment_factory` is not a supported use case for this PR— the two features address independent needs and their interaction is out of scope here.

[the remark this answers is not in the document]
Comment (lailanelkoussy) [reply trl/trainer/grpo_trainer.py:1838]: Adding documentation to explain this choice

File: tests/test_grpo_tools_column.py [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+            trainer.train()
+
+        assert trainer.state.log_history[-1]["train_loss"] is not None
+        assert trainer.state.log_history[-1]["tools/call_frequency"] == pytest.approx(2 / 3)
+        assert trainer.state.log_history[-1]["tools/failure_frequency"] == pytest.approx(0.0)
+        # Verify correct per-environment binding: each environment's own increment was called.
+        # If all calls resolved to environments[0] (old registry fallback), then
+        # environments[0]._counter would be 2 and environments[1]._counter would be 0.
+        assert trainer.environments[0]._counter == 1
+        assert trainer.environments[1]._counter == 1
```
[the remark this answers is not in the document]
Comment (lailanelkoussy) [reply tests/test_grpo_tools_column.py]: Indeed, the test assertions were wrong. All tool calls resolve to environment 0's bound methods since per_sample_tools is built from
  self.tools (which only holds env 0's callables). Fixed: the test now asserts environments[0]._counter == 2 and environments[1]._counter == 0,
  matching the documented limitation.

**Comment by lailanelkoussy** [comment]
Thanks for the review @qgallouedec! Here's a summary of the main changes since I made your feedback:

**Addressing your comments:**
- **Removed `tools_column_name` parameter entirely** — the trainer now auto-detects a `"tools"` column in the dataset, no new init arg needed.
- **Removed duplicate tool name validation** — agreed it was unnecessary defensiveness.
- **Removed all dataset-level validation** (the init-time iteration over dataset rows checking for unknown tool names) — too defensive per your feedback.
- `"tools"` is now unconditionally included in `_signature_columns` so the column passes through the dataloader when present.

**`tools` column + `environment_factory` — intentional non-support:**

When `environment_factory` is set, the trainer creates N environment instances (one per `generation_batch_size` slot) and binds each environment's methods into per-slot tool dicts (`_sync_tool_dicts[i]` / `_async_tool_dicts[i]`) so that sample `i` executes on environment `i`. However, `self.tools` only stores environment 0's bound methods (line 491: `tools + environment_methods[0]`), because `self.tools` is used for schema/tokenization purposes and all environments share the same method signatures.

Per-sample tool filtering builds its lookup from `self.tools`, which means all resolved callables point to environment 0's instance. Making this work correctly would require cross-referencing the dataset `tools` column with the per-slot `_sync_tool_dicts[i]` / `_async_tool_dicts[i]`, coupling the filtering logic tightly to the environment lifecycle. This adds real complexity for a use case that doesn't have a clear practical motivation yet — when using environments, the environment itself typically controls which tools are available.

Rather than adding that complexity now, I chose to document this as a known limitation (comment in `_generate_and_score_completions`) and write a test that verifies training still completes without errors, asserting the actual behavior (all calls dispatch to environment 0) rather than the ideal behavior.


Let me know if you agree with this approach or if you'd rather I do things differently. Thanks in advance
